wrappers/rust/cpp-client-telemetry-sys/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+use std::ffi::*;
+
+fn evt_api_call_wrapper(evt_context: Box<evt_context_t>) -> (evt_status_t, Box<evt_context_t>) {
+    let raw_pointer = Box::into_raw(evt_context);
+
+    let mut result: evt_status_t = -1;
+    unsafe {
+        result = evt_api_call_default(raw_pointer);
+    }
+
+    let out_context = unsafe { Box::from_raw(raw_pointer) };
+    (result, out_context)
+}
+
+pub fn evt_open(config: &CString) -> Option<evt_handle_t> {
+    let config_bytes = config.to_bytes_with_nul().to_vec();
+
+    let context: Box<evt_context_t> = Box::new(evt_context_t {
+        call: evt_call_t_EVT_OP_OPEN,
+        handle: 0,
+        data: config_bytes.as_ptr() as *mut c_void,
+        result: 0,
+        size: 0,
+    });
+
+    let (result, context) = evt_api_call_wrapper(context);
+
+    if result == -1 {
+        return Option::None;
+    }
+
+    return Some(context.handle.clone());
+}
+
+pub fn evt_close(handle: &evt_handle_t) -> evt_status_t {
+    let context: Box<evt_context_t> = Box::new(evt_context_t {
+        call: evt_call_t_EVT_OP_CLOSE,
+        handle: *handle,
+        data: std::ptr::null_mut(),
+        result: 0,
+        size: 0,
+    });
+
+    let (result, _) = evt_api_call_wrapper(context);
+    result
+}
+
+pub fn evt_upload(handle: &evt_handle_t) -> evt_status_t {
+    let context: Box<evt_context_t> = Box::new(evt_context_t {
+        call: evt_call_t_EVT_OP_UPLOAD,
+        handle: *handle,
+        data: std::ptr::null_mut(),
+        result: 0,
+        size: 0,
+    });
+
+    evt_api_call_wrapper(context).0
+}
+
+pub fn evt_flush(handle: &evt_handle_t) -> evt_status_t {
+    let context: Box<evt_context_t> = Box::new(evt_context_t {
+        call: evt_call_t_EVT_OP_FLUSH,
+        handle: *handle,
+        data: std::ptr::null_mut(),
+        result: 0,
+        size: 0,
+    });
+
+    evt_api_call_wrapper(context).0
+}
+
+pub fn evt_log(handle: &evt_handle_t, data: &mut [evt_prop]) -> evt_status_t {
```

Comment:
This function must have an unsafe signature, i.e. `pub unsafe fn evt_log`, because the values inside `evt_prop` use raw pointers and the implementation of this function dereferences those raw pointers.  As written now, this function could be used to violate type safety.

A better solution could ensure that `evt_prop` was using type-safe types.  You could have a type which represents an owned pointer to a C-style string, that has the same representation as the underlying `evt_*` structure.  That would greatly reduce your exposure to unsafe code semantics.

If you can point out where the C headers are that define those structures, I can help write the safe Rust projections for these.